File: snippets/java/src/main/java/develop/Signals.java
package develop;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import java.time.Duration;

// Demo app for the signals docs examples — run against a local Restate server on port 9084.
@Service
public class Signals {

  public record SteerRequest(String invocationId, String text) {}

  public record ApproveRequest(String invocationId, boolean approved) {}

  // <start_one_shot>
  @Handler
  public Boolean waitForApproval() {
    return Restate.signal("approval", Boolean.class).await();
  }

  // <end_one_shot>

  // <start_wait>
  @Handler
  public String reviseUntilDone(String topic) {
    String draft = "Research notes for " + topic;
    while (true) {
      String text = Restate.signal("steer", String.class).await();
      if (text.equals("done")) {
        return draft;
      }
      draft = draft + "\n" + text;
    }
  }

  // <end_wait>

  @Handler
  public String sleepThenWait() {
    Restate.sleep(Duration.ofSeconds(3));
    return Restate.signal("steer", String.class).await();
  }

  // <start_resolve>
  @Handler
  public void steerInvocation(SteerRequest req) {
    Restate.invocationHandle(req.invocationId()).signal("steer").resolve(String.class, req.text());
  }

  // <end_resolve>

  @Handler
  public void approve(ApproveRequest req) {
    Restate.invocationHandle(req.invocationId())
        .signal("approval")
        .resolve(Boolean.class, req.approved());
  }

  @Handler
  public void deny(SteerRequest req) {
    Restate.invocationHandle(req.invocationId()).signal("approval").reject("Request denied");
  }
}

File: snippets/kotlin/src/main/kotlin/develop/Signals.kt
package develop

// Demo app for the signals docs examples — run against a local Restate server on port 9085.
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.*
import kotlin.time.Duration.Companion.seconds
import kotlinx.serialization.Serializable

@Serializable data class SteerRequest(val invocationId: String, val text: String)

@Serializable data class ApproveRequest(val invocationId: String, val approved: Boolean)

@Service
class Signals {

  @Handler
  suspend fun waitForApproval(): Boolean {
    // <start_one_shot>
    var approval = signal<Boolean>("approval").await()
    // <end_one_shot>
    return approval
  }


  // <start_wait>
  @Handler
  suspend fun reviseUntilDone(topic: String): String {
    var draft = "Research notes for $topic"
    while (true) {
      val text = signal<String>("steer").await()
      if (text == "done") {
        return draft
      }
      draft = "$draft\n$text"
    }
  }

  // <end_wait>

  @Handler
  suspend fun sleepThenWait(): String {
    sleep(3.seconds)
    return signal<String>("steer").await()
  }

  @Handler
  suspend fun steerInvocation(req: SteerRequest) {
    // <start_resolve>
    invocationHandle<Unit>(req.invocationId).signal("steer").resolve(req.text)
    // <end_resolve>
  }


  @Handler
  suspend fun approve(req: ApproveRequest) {
    invocationHandle<Unit>(req.invocationId).signal("approval").resolve(req.approved)
  }

  @Handler
  suspend fun deny(req: SteerRequest) {
    invocationHandle<Unit>(req.invocationId).signal("approval").reject("Request denied")
  }
}
